File: src/main/java/com/hjf/java/spark/code/Test05.java
package com.hjf.java.spark.code;

import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.VoidFunction;
import org.codehaus.janino.Java;
import scala.Tuple2;
import scala.collection.mutable.ListBuffer;


import java.util.*;

/**
 * @author Jiang锋时刻
 * @create 2020-04-05 22:14
 */
public class Test05 {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setMaster("local").setAppName("Test05");
        JavaSparkContext sc = new JavaSparkContext(conf);
//        JavaRDD<String> rdd1 = sc.parallelize(Arrays.asList(
//                "str1", "str2", "str3", "str4",
//                "str5", "str6", "str7", "str8",
//                "str9", "str10", "str11", "str12"
//        ), 3);
//        JavaRDD<String> rdd2 = rdd1.mapPartitionsWithIndex(new Function2<Integer, Iterator<String>, Iterator<String>>() {
//            @Override
//            public Iterator<String> call(Integer index, Iterator<String> iter) throws Exception {
//                List<String> list = new ArrayList<>();
//                while (iter.hasNext()) {
//                    String one = iter.next();
//                    list.add("rdd1 partition = [" + index + "], values = [" + one + "]");
//                }
//                return list.iterator();
//            }
//        }, true);
////        List<String> result = rdd2.collect();
////
////        for (String s : result){
////            System.out.println(s);
////        }
//
//         /*
//          repartition, coalesce
//         */
//        JavaRDD<String> rdd3 = rdd2.repartition(4);
//        JavaRDD<String> rdd4 = rdd3.mapPartitionsWithIndex(new Function2<Integer, Iterator<String>, Iterator<String>>() {
//            @Override
//            public Iterator<String> call(Integer index, Iterator<String> iter) throws Exception {
//                List<String> list = new ArrayList<>();
//                while (iter.hasNext()) {
//                    String one = iter.next();
//                    list.add("rdd3 partition = [" + index + "], values = [" + one + "]");
//                }
//                return list.iterator();
//            }
//        }, true);
//        List<String> result = rdd4.collect();
//        for (String s : result){
//            System.out.println(s);
//        }



//        JavaRDD<String> rdd1 = sc.parallelize(Arrays.asList("张三", "李四", "王五", "张三"));
//        JavaRDD<String> rdd2 = sc.parallelize(Arrays.asList("张三", "李四", "田七", "周八"));
        /*
            zip
         */
//         JavaPairRDD<String, String> rdd3 = rdd1.zip(rdd2);
//         List<Tuple2<String, String>> result = rdd3.collect();
//         for(Tuple2<String, String> s : result){
//             System.out.println(s);
//         }

        /*
            zipWithIndex
         */
//        JavaPairRDD<String, Long> rdd4 = rdd1.zipWithIndex();
//        List<Tuple2<String, Long>> result = rdd4.collect();
//        for (Tuple2<String, Long> s :result){
//            System.out.println(s);
//        }

        /*
            groupByKey
         */
//        JavaPairRDD<String, Long> rdd4 = rdd1.zipWithIndex();
//        JavaPairRDD<String, Iterable<Long>> rdd5 = rdd4.groupByKey();
//        List<Tuple2<String, Iterable<Long>>> result = rdd5.collect();
//        for(Tuple2<String, Iterable<Long>> s : result){
//            System.out.println(s);
//        }


        /*
            reduce
         */
//        JavaRDD<Integer> rdd1 = sc.parallelize(Arrays.asList(1, 2, 3, 4, 5));
//        Integer result = rdd1.reduce(new Function2<Integer, Integer, Integer>() {
//            @Override
//            public Integer call(Integer v1, Integer v2) throws Exception {
//                return v1 + v2;
//            }
//        });
//        System.out.println(result);


        JavaPairRDD<String, Integer> rdd1 = sc.parallelizePairs(Arrays.asList(new Tuple2("张三", 10),
            new Tuple2("张三", 10), new Tuple2("李四", 20), new Tuple2("李四", 20), new Tuple2("王五", 30),
            new Tuple2("王五", 300), new Tuple2("赵六", 40), new Tuple2("赵六", 400)
        ));

//        Map<String, Long> map = rdd1.countByKey();
//        Set<Map.Entry<String, Long>> entries = map.entrySet();
//        for(Map.Entry<String, Long> entry : entries){
//            String key = entry.getKey();
//            Long value = entry.getValue();
//
//            System.out.println("key = " + key + ", value = " + value);
//
//        }
        Map<Tuple2<String, Integer>, Long> map = rdd1.countByValue();
        Set<Map.Entry<Tuple2<String, Integer>, Long>> entries = map.entrySet();
        for(Map.Entry<Tuple2<String, Integer>, Long> entry : entries){
            Tuple2<String, Integer> key = entry.getKey();
            Long value = entry.getValue();

            System.out.println("key = " + key + ", value = " + value);

        }


    }
}

File: src/main/scala/com/hjf/scare/spark/core/Test05.scala
package com.hjf.scare.spark.core

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * @author Jiang锋时刻
 * @create 2020-04-05 20:03
 */
object Test05 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("Test05").setMaster("local")
    val sc: SparkContext = new SparkContext(conf)

    /*
      countByKey:统计key-value对中相同key的数量
      countByValue:将原数据对当做key值,统计各key值(原数据)的数量
     */

//    val rdd1: RDD[(String, Int)] = sc.parallelize(List[(String, Int)](
//      ("a", 100), ("b", 200), ("c", 300), ("a", 100), ("b", 2), ("c", 300)))
//    val count: collection.Map[String, Long] = rdd1.countByKey()
//    count.foreach(println)
//    /*
//      (a,2)
//      (b,2)
//      (c,2)
//     */

//    val values: collection.Map[(String, Int), Long] = rdd1.countByValue()
//    values.foreach(println)
//    /*
//      ((b,2),1)
//      ((a,100),2)
//      ((c,300),2)
//      ((b,200),1)
//     */

    // 不一定是key-value格式
    val rdd2: RDD[String] = sc.parallelize(List[String]("hello", "world", "hive", "hbase", "hive", "world"))
    val result: collection.Map[String, Long] = rdd2.countByValue()
    result.foreach(println)
    /*
      (hive,2)
      (hello,1)
      (hbase,1)
      (world,2)
     */






    /*
      reduce
     */
//    val rdd1: RDD[Int] = sc.parallelize(List[Int](1, 2, 3, 4, 5))
//    val result: Int = rdd1.reduce((v1, v2) => {v1 + v2})
//    // val result: Int = rdd1.reduce(_ + _)
//    println(result)



    /*
      groupByKey:根据key值聚合
     */
//    val rdd1: RDD[(String, Int)] = sc.parallelize(List[(String, Int)](("a", 10), ("a", 20), ("b", 30), ("c", 40)))
//    val result: RDD[(String, Iterable[Int])] = rdd1.groupByKey()
//    result.foreach(println)
//    /*
//      (a,CompactBuffer(10, 20))
//      (b,CompactBuffer(30))
//      (c,CompactBuffer(40))
//     */


    /*
      zip, zipWithIndex
     */
//    val rdd1: RDD[String] = sc.parallelize(List[String]("张三", "李四", "王五", "赵六"))
//    val rdd2: RDD[String] = sc.parallelize(List[String]("100", "200", "300", "400"))
//    // zip:将两个数据集压缩在一起,但必须保证两个数据集中数据量必须保证一致,否则会报错
//     val result: RDD[(String, String)] = rdd1.zip(rdd2)
//     result.foreach(println)
//    /*
//      (张三,100)
//      (李四,200)
//      (王五,300)
//      (赵六,400)
//     */
//
//    // zipWithIndex:将数据集中的内容和其在数据集中下标组成key-value压缩
////    val result: RDD[(String, Long)] = rdd1.zipWithIndex()
////    result.foreach(println)
//    /*
//      (张三,0)
//      (李四,1)
//      (王五,2)
//      (赵六,3)
//     */

    /*
      repartition, coalesce
     */
//    val rdd1: RDD[String] = sc.parallelize(List[String](
//      "str1", "str2", "str3", "str4",
//      "str5", "str6", "str7", "str8",
//      "str9", "str10", "str11", "str12"
//    ), 3)
//    // index: 分区号
//    // iter:分区号下的数据
//    val rdd2: RDD[String] = rdd1.mapPartitionsWithIndex((index, iter) => {
//      val list = new ListBuffer[String]()
//      while (iter.hasNext) {
//        val one: String = iter.next()
//        list.+=(s"rdd1 partition = [$index], values = [$one]")
//      }
//      list.iterator
//    })
//    //rdd2.foreach(println)
//
//    // repartition:可以修改分区数,默认会产生shuffle  [宽依赖]
//    // coalesce:可以修改分区数,默认不会产生shuffle   [窄依赖]
//    // coalesce(2, true):就相当于repartition.repartition的底层就是coalesce(2, true)
//
//    // 注意:当coalesce由少的分区分到多的分区时,不让产生suffer,不起作用.
//    // repartition常用于增多分区
//    // coalesce常用于减少分区
//    val rdd3: RDD[String] = rdd2.coalesce(4)
//    val rdd4: RDD[String] = rdd3.mapPartitionsWithIndex((index, iter) => {
//      val list = new ListBuffer[String]()
//      while (iter.hasNext) {
//        val one: String = iter.next()
//        list.+=(s"rdd1 partition = [$index], values = [$one]")
//      }
//      list.iterator
//    })
//
//    //rdd4.foreach(println)
//
//    val result: Array[String] = rdd4.collect()
//    result.foreach(println)

  }

}
